### PR TITLE
Role entity and ability access to controller endpoints by role added

### PR DESCRIPTION
**Pull request by maiiat**
1. spring-boot-starter-security added
2. SecurityConfig class with PasswordEncoder and SecurityFilterChain beans added
3. UserDetails interface in the User class implemented
4. CustomUserDetailsService class with UserDetailsService interface added. loadUserByUsername method overrided
5. Role/RoleRepository added
6. Set<Role> roles field added to the User entity
7. 03-create-roles-table.yaml, 04-insert-default-roles.yaml, 05-create-users-roles-table.yaml, 06-assign-default-role-to-users.yaml, 07-upsert-admin-user.yaml files added to databaseChangeLog
8. @PreAuthorize with proper role access over the existing BookController endpoints added
9. BookServiceImpl class refactored with creation new findBookByIdOrThrowException() method
10. CreateUpdateBookRequestDto & UserRegistrationRequestDto records extended with additional annotation that better reflects entities restriction in Swagger
11. Isbn & Password & PasswordConstraintValidator refactored
12. JOIN FETCH added to UserRepository.findByEmail to avoid LazyInitializationException
13. PasswordEncoder added to UserServiceImpl.register

### CODE REVIEW

File: src/main/java/com/example/service/impl/BookServiceImpl.java [hunk 1]
```diff
@@ -75,7 +73,12 @@ public List<BookDto> search(BookSearchParameters bookSearchParameters) {
 
     private void checkIfIsbnExistsOrThrowException(String isbn) {
         if (bookRepository.existsByIsbn(isbn)) {
-            throw new IsbnAlreadyExistException("such ISBN already exist: " + isbn);
+            throw new IsbnAlreadyExistException("Such ISBN already exist: " + isbn);
         }
     }
+
+    private Book findBookByIdOrThrowException(Long id) {
```
Comment (MentorMA) [inline src/main/java/com/example/service/impl/BookServiceImpl.java:80]:
```suggestion
    private Book findBookById(Long id) {
```

File: src/main/java/com/example/service/impl/BookServiceImpl.java [hunk 2]
```diff
@@ -75,7 +73,12 @@ public List<BookDto> search(BookSearchParameters bookSearchParameters) {
 
     private void checkIfIsbnExistsOrThrowException(String isbn) {
```
Comment (MentorMA) [inline src/main/java/com/example/service/impl/BookServiceImpl.java:74]:
```suggestion
    private void checkIfIsbnExists(String isbn) {
```

File: src/main/java/com/example/service/impl/UserServiceImpl.java [hunk 3]
```diff
@@ -8,25 +8,32 @@
...
+        user.setPassword(passwordEncoder.encode(userRegistrationRequestDto.password()));
+        user.setEmail(userRegistrationRequestDto.email());
+        user.setFirstName(userRegistrationRequestDto.firstName());
+        user.setLastName(userRegistrationRequestDto.lastName());
+        user.setShippingAddress(userRegistrationRequestDto.shippingAddress());
         User savedUser = userRepository.save(user);
         return userMapper.toDto(savedUser);
     }
 
     private void checkIfEmailExistsOrThrowException(String email) {
```
Comment (MentorMA) [inline src/main/java/com/example/service/impl/UserServiceImpl.java:34]:
```suggestion
    private void checkIfEmailExists(String email) {
```

File: src/main/java/com/example/service/impl/UserServiceImpl.java [hunk 4]
```diff
@@ -8,25 +8,32 @@
...
 
 @RequiredArgsConstructor
 @Service
 public class UserServiceImpl implements UserService {
     private final UserRepository userRepository;
     private final UserMapper userMapper;
+    private final PasswordEncoder passwordEncoder;
 
     @Override
     public UserDto register(UserRegistrationRequestDto userRegistrationRequestDto) {
```
Comment (MentorMA) [inline src/main/java/com/example/service/impl/UserServiceImpl.java:22]:
```suggestion
    public UserDto register(UserRegistrationRequestDto user) {
```

Comment (MentorMA) [reply src/main/java/com/example/service/impl/UserServiceImpl.java:22]: Here you can simplify

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/controller/BookController.java [hunk 5]
```diff
@@ -29,24 +30,28 @@ public class BookController {
     private final BookService bookService;
 
     @GetMapping
+    @PreAuthorize("hasRole('USER')")
```
Comment (olekskov) [inline src/main/java/com/example/controller/BookController.java:33]: In such way you give no access to endpoint to Admin, please recheck such behaviour is intended

**Review by olekskov — APPROVED** [review]

File: src/main/java/com/example/service/impl/BookServiceImpl.java (hunk 1 above)
Comment (maiiat) [reply src/main/java/com/example/service/impl/BookServiceImpl.java:80]: done

File: src/main/java/com/example/service/impl/UserServiceImpl.java (hunk 3 above)
Comment (maiiat) [reply src/main/java/com/example/service/impl/UserServiceImpl.java:34]: done

File: src/main/java/com/example/service/impl/UserServiceImpl.java (hunk 4 above)
Comment (maiiat) [reply src/main/java/com/example/service/impl/UserServiceImpl.java:22]: done

File: src/main/java/com/example/controller/BookController.java (hunk 5 above)
Comment (maiiat) [reply src/main/java/com/example/controller/BookController.java:33]: Its also strange for me, but according to requirements here shouldn't be get method accessable:
image

File: src/main/java/com/example/service/impl/BookServiceImpl.java (hunk 2 above)
Comment (maiiat) [reply src/main/java/com/example/service/impl/BookServiceImpl.java:74]: done

**Pull request merged**